File: LeetCode/Easy/0101-symmetric-tree/0101-symmetric-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private const int MaxNodes = 1000;
    public bool IsSymmetric(TreeNode root) {
        var stack = new Stack<(TreeNode, TreeNode)>(MaxNodes / 2);

        stack.Push((root.left, root.right));
        
        while (stack.Count > 0)
        {
            var (lNode, rNode) = stack.Pop();

            if (lNode == null && rNode == null) continue;   // symmetry - continue
            if (lNode == null || rNode == null || lNode.val != rNode.val) return false; // tree or value asymmetry 
            
            stack.Push((lNode.left, rNode.right));
            stack.Push((lNode.right, rNode.left));
        }

        return true;
    }
}

File: LeetCode/Easy/0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        stack: List[Tuple[TreeNode, TreeNode]] = []
        stack.append((root.left, root.right))

        while stack:
            left, right = stack.pop()
            if left is None and right is None:
                continue
            if left is None or right is None or left.val != right.val:
                return False
            
            stack.append((left.right, right.left))
            stack.append((left.left, right.right))
        
        return True